Branch-SDK/src/main/java/io/branch/referral/AttributionReportingManager.kt

```diff
@@ -0,0 +1,166 @@
+
+    fun registerSource(context: Context) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            if (SdkExtensions.getExtensionVersion(SdkExtensions.AD_SERVICES) >= 7) {
+                if (isMeasurementApiEnabled()) {
+                    val manager = MeasurementManager.get(context)
+                    val executor = Executors.newSingleThreadExecutor()
+                    val params = getParams(context)
+                    val branchBaseURL = PrefHelper.getInstance(context).apiBaseUrl
+                    val sourceUri = Uri.parse("${branchBaseURL}${Defines.RequestPath.RegisterSource}?$params")
+
+                    manager.registerSource(sourceUri, null, executor, object : OutcomeReceiver<Any?, Exception> {
+                        override fun onResult(result: Any?) {
+                            BranchLogger.v("Source registered successfully with URI: $sourceUri")
+                            executor.shutdown()
+                        }
+
+                        override fun onError(e: Exception) {
+                            BranchLogger.w("Error while registering source: ${e.message}")
+                            executor.shutdown()
+                        }
+                    })
+                }
+            }
+        }
+    }
+
+    fun registerTrigger(context: Context, conversionId: String) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            if (SdkExtensions.getExtensionVersion(SdkExtensions.AD_SERVICES) >= 7) {
+                if (isMeasurementApiEnabled()) {
+                    BranchLogger.v("Registering trigger for conversion ID: $conversionId")
+
+                    val executor = Executors.newSingleThreadExecutor()
+                    val params = getParams(context) + "&conversion_id=$conversionId"
+                    val branchBaseURL = PrefHelper.getInstance(context).apiBaseUrl
+                    val triggerUri = Uri.parse("${branchBaseURL}${Defines.RequestPath.RegisterTrigger}?$params")
+                    val manager = MeasurementManager.get(context)
+
+                    manager.registerTrigger(triggerUri, executor, object : OutcomeReceiver<Any?, Exception> {
+                            override fun onResult(result: Any?) {
+                                BranchLogger.v("Trigger registered successfully with URI: $triggerUri")
+                                executor.shutdown()
+                            }
+
+                            override fun onError(e: Exception) {
+                                BranchLogger.w("Error while registering trigger: ${e.message}")
+                                executor.shutdown()
+                            }
+                        }
+                    )
+                } else {
+                    BranchLogger.v("Measurement API is not enabled. Did not register trigger.")
+                }
+            }
+        }
+    }
+
+    fun getParams(context: Context): String {
```

Comment:
What's required and what's optional? This is a pretty big list, and will very easily get impacted by any device info changes.

Reply:
We want to include them all, even if the values are null or empty